### DISCUSSION TITLE
Optimize Pdfium image rendering

### DISCUSSION THREAD

**Discussion by swftvsn** [libvips/libvips#2009] [Ideas]
The relevant discussion is in #689, see: https://github.com/jcupitt/libvips/issues/689#issuecomment-385024057 

Copied here too:

> hey @jcupitt
> 
> found some information around the rendering implementation by using FPDF_RenderPageBitmapWithMatrix it may help keep memory a little lower by implementing a tiling load
> 
> [fpdf_view_embeddertest.cpp#403](https://pdfium.googlesource.com/pdfium/+/master/fpdfsdk/fpdf_view_embeddertest.cpp#403)
> 
> Render the (2, 1) tile of the page (third column, second row) when the page is divided in 50x50 pixel tiles. The tile is scaled by a factor of 7.
> 
> [fpdf_view_embeddertest.cpp#557](https://pdfium.googlesource.com/pdfium/+/master/fpdfsdk/fpdf_view_embeddertest.cpp#557)
> 
> also i wonder if there is some performance gains by decreasing the bitmap formatter a little by using FPDFBitmap_CreateEx(height, width, FXDIB_8bppRgb)
> 
> [fpdf_view.cpp#810](https://pdfium.googlesource.com/pdfium/+/master/fpdfsdk/fpdf_view.cpp#810)
> 
> let me know your thoughts.

**Comment by jcupitt** [comment]
Just a quick pdfium update: the way we were packaging and linking to libpdfium before has changed (the project still seems to be in very heavy flux) and pdfload via pdfium is not currently building.

I spent a day trying to bring it back to life, but it seems git master pdfium simply does not support building as a library. There are various issues open on this, hopefully it'll start working again soon.

**Comment by SPhu** [comment]
I have been having some mixed performance issues with poppler compared to imagick/ghostscript for converting pdfs to images.   Poppler seems to be much slower on small pdf files (3MB, gs=.15s, poppler=1.5s).   gs is much slower on large pdfs (500MB, gs=21s, poppler=6s). I tried getting pdfium to work with libvips and failed, but it sounds like its not just me.  

@jcupitt  

1. what are your thoughts on using forked binaries like this for pdfium?  They include some patches that look like they are an improvement for including in other projects.  https://github.com/bblanchon/pdfium-binaries 

2. mupdf seems to give similar output and similar speed to pdfium but seems to have better packaging and distribution support.  Is mupdf something you already looked at? 

  
The current workaround I am using is preprocessing the file with mupdf's mutool convert cli tool to convert from pdf to png first.  This gives me the speed boost I'm looking for but inflates my container size and means I have to separately execute multiple processes.  Are there any better solutions you recommend?

**Comment by jcupitt** [comment]
Hi @SPhu,

1. It sounds like it might be even harder to support :( Hopefully pdfium will settle down and become simpler to work with, and I'd rather wait for that.
2. mupdf is very nice, but the licence is impossible to work with, unfortunately. libvips can't use GhostScript for the same reason.

Perhaps another solution would be to revert to the version of pdfium which was working way back in 2017 or whenever it was.
